### PR TITLE
Issue 16 - Utility 3 - Add theme clicking-preview functionality

### PR DESCRIPTION
**Pull request by FrancescoSTL**
This PR adds the switch case for the theme button clicks, as well as the function needed to change the user's theme.

Issue ref: #16

### CODE REVIEW

File: data/js/utility/content3.js [hunk 1]
```diff
@@ -1,12 +1,10 @@
...
-    var timeout;
-    // set template and hoveremit names
+    // set template names
     var templateNames = ['template1','template2','template3'];
-    var hoverEmit = ['hoverTemplate1','hoverTemplate2','hoverTemplate3'];
 
-    // add listeners to all of our buttons for hovering and clicking
+    // add listeners to all of our buttons for clicking
     for(var x=0; x<templateNames.length; x++)
     {
```
Comment (schalkneethling) [inline data/js/utility/content3.js]: Nit: For consistent code style do `for (var x = 0....) {` i.e. opening curly brace on same line as `for` statement. Also, cache the array length. This improves performance and is a general best practice:

```
for (var i = 0, l = templateNames.length; i < l; i++) {
```

File: index.js [hunk 2]
```diff
@@ -166,6 +166,56 @@ function showBadge() {
 }
 
 /**
+ * Purpose: Changes the theme based upon the value passed
```
Comment (schalkneethling) [inline index.js]: Nit: Again here, "Purpose" is implied.

File: index.js [hunk 3]
```diff
@@ -166,6 +166,56 @@ function showBadge() {
...
+ * @param {int}     themeNum - a number passed based upon what theme button the user selected
```
Comment (schalkneethling) [inline index.js]: Nit: spacing after `{int}`
Perhaps the above can be rephrased to just: "based on the theme button clicked by the user" ?

File: index.js [hunk 4]
```diff
@@ -166,6 +166,56 @@ function showBadge() {
...
+ */
+function changeTheme(themeNum) {
+    // set the theme slug based upon the number theme passed
```
Comment (schalkneethling) [inline index.js]: "based on the theme number passed"?

File: index.js [hunk 5]
```diff
@@ -166,6 +166,56 @@ function showBadge() {
...
+    if(themeNum == 1)
```
Comment (schalkneethling) [inline index.js]: Always use `===`

File: index.js [hunk 6]
```diff
@@ -166,6 +166,56 @@ function showBadge() {
...
+        var personaSlug = "fox-in-snow";
```
Comment (schalkneethling) [inline index.js]: Always use single quotes

File: index.js [hunk 7]
```diff
@@ -166,6 +166,56 @@ function showBadge() {
...
+    else if(themeNum == 2)
+        var personaSlug = "cozy-fox";
+    else if(themeNum == 3)
+        var personaSlug = "shiretoko-fox";
```
Comment (schalkneethling) [inline index.js]: You do not need to declare the variable each time instead, declare it once as the first line of the function, and then simply update it here. So,

```
function changeTheme(themeNum) {
    var personaSlug;
    ....
```

File: index.js [hunk 8]
```diff
@@ -166,6 +166,56 @@ function showBadge() {
...
+    // if there is no number passed, set the theme to default and return
+    else
+    {
```
Comment (schalkneethling) [inline index.js]: Coding style:
```
else {
....
}
```

File: index.js [hunk 9]
```diff
@@ -166,6 +166,56 @@ function showBadge() {
...
+        LightweightThemeManager.themeChanged(null);
+        return;
+    }
+
+    // start a new XMLHTTP request with AMO to request the page where we can get the addon ID
+    var request = new XMLHttpRequest();
+    request.open("GET", "https://services.addons.mozilla.org/firefox/api/addon/" + personaSlug);
```
Comment (schalkneethling) [inline index.js]: Always use single quotes.

File: index.js [hunk 10]
```diff
@@ -166,6 +166,56 @@ function showBadge() {
...
+    request.onload = function() {
+        try {
+            // get the addon ID from the page returned
+            var id = request.responseXML.documentElement.getAttribute("id");
+
+            // start a new XMLHTTP request to get the theme JSON from AMO
+            var personaRequest = new XMLHttpRequest();
+            personaRequest.open("GET", "https://versioncheck.addons.mozilla.org/en-US/themes/update-check/" + id);
```
Comment (schalkneethling) [inline index.js]: Do we really need two requests? As we are using a specific set of themes, can we not simply hard code the theme ids as an array.

Above where you set the theme slug, you can set the theme id or, even skip that, and simply use the `themeNum` number as the array index. Then we only need this one request.

File: index.js [hunk 11]
```diff
@@ -166,6 +166,56 @@ function showBadge() {
...
+            personaRequest.onload = function() {
+                try {
+                    // get the theme JSON from the response
+                    var theme = JSON.parse(personaRequest.response);
+                    // set the theme
+                    LightweightThemeManager.themeChanged(theme);
+                }
+                catch (e) {
+                    showErrorMessage("Invalid Persona");
```
Comment (schalkneethling) [inline index.js]: Where is `showErrorMessage` defined? Also, you are taking in `e` but not using it.

File: index.js [hunk 12]
```diff
@@ -200,6 +250,19 @@ function showSidebar(sidebarProps, contentURL) {
...
+                    case 'template1':
+                        changeTheme(1);
+                        break;
+                    case 'template2':
+                        changeTheme(2);
+                        break;
+                    case 'template3':
+                        changeTheme(3);
+                        break;
+                        break;
```
Comment (schalkneethling) [inline index.js]: Duplicate `break;`

**Comment by FrancescoSTL** [comment]
Revised and ready for review

File: index.js [hunk 13]
```diff
@@ -202,6 +202,45 @@ function highLight(item) {
...
+        personaSlug = 'fox-in-snow';
+    else if(themeNum === 2)
+        personaSlug = 'cozy-fox';
+    else if(themeNum === 3)
+        personaSlug = 'shiretoko-fox';
+    // if there is no number passed, set the theme to default and return
+    else {
+        LightweightThemeManager.themeChanged(null);
+        return;
+    }
```
Comment (schalkneethling) [inline index.js]: I guess we not need the `if.. else..` anymore?

Comment (FrancescoSTL) [inline index.js]: Nope, I didn't catch that. I'll still need to check if themeNum is null so that I can reset the theme.

Comment (schalkneethling) [inline index.js]: Basically
```
if (typeof themeNum === 'undefined') {
   return;
}
...
```

Comment (FrancescoSTL) [inline index.js]: I suppose I'll have to replace `changeTheme(null);` with `changeTheme();` if we deal with undefined rather than null

**Comment by FrancescoSTL** [comment]
Done and done 👍

**Comment by FrancescoSTL** [comment]
Just realized I didn't remove my toggleSidebar() call. Whoops.

**Comment by FrancescoSTL** [comment]
*Now* all is done 👊

**Pull request merged**